File: QUICKSTARTS/f#/MyStrategy.fs
namespace FSharpCgdk

open FSharpCgdk.Model
open FSharp.Json

// Данные, в которых будем хранить атакующего бота, чтобы не пересчитывать дважды
// forTick - для какого тика актуальна информация, botId - id атакующего бота
type StrategyData = KnownAttackBot of forTick: int * botId : int 


module MyStrategy = 
    // Эпсилон! :D
    let private EPS = 0.1

    // Краткое описание собственной веткорной алгебры.
    type Vector = V3 of X : float * Y : float * Z : float

    let vX (V3(x, y, z)) = x
    let vY (V3(x, y, z)) = y
    let vZ (V3(x, y, z)) = z

    let add (V3(x1, y1, z1)) (V3(x2, y2, z2)) = 
        V3(x1 + x2, y1 + y2, z1 + z2)

    let sub (V3(x1, y1, z1)) (V3(x2, y2, z2)) = 
        V3(x1 - x2, y1 - y2, z1 - z2)

    let mulScalar (V3(x, y, z)) mul = V3(x * mul, y * mul, z * mul)
    let divScalar vector div = mulScalar vector (1.0/div)

    let length (V3(x, y, z)) = sqrt (x * x + y * y + z * z)

    let projectionXZ (V3(x, y, z)) = V3(x, 0.0, z)

    let positionBall (ball : Ball) = V3(ball.x, ball.y, ball.z)    
    let positionRobot (robot : Robot) = V3(robot.x, robot.y, robot.z)   

    let velocityBall (ball : Ball) = V3(ball.velocity_x, ball.velocity_y, ball.velocity_z)    
    let velocityRobot (robot : Robot) = V3(robot.velocity_x, robot.velocity_y, robot.velocity_z)   


    // Та самая хранимая информация.
    // Инициализируем не существующим тиком.
    let mutable data : StrategyData = KnownAttackBot(-1, -1)

    // Дальше следует читать программу снизу вверх для лучшего понимания
    // Последовательность чтения:
    // act -> nextData ->  attackAct -> entryAttackPoint -> defeatAct


    let protectAct (me, rules, game, _) : Action =

        // Вычисляем куда встать как вратари на проекции оси OZ
        let target_pos_z = -(rules.arena.depth / 2.0) + rules.arena.bottom_radius
        // Вычисляем куда встать на проекции OX
        let time = (target_pos_z - game.ball.z) / game.ball.velocity_z  
        let x = game.ball.x + game.ball.velocity_x * time

        // Сама точка перехвата
        let interception = 
            if game.ball.velocity_z > -EPS then 
                V3(0.0, 0.0, target_pos_z)
            else
                V3(x, 0.0, target_pos_z)

                
        let max_speed = rules.ROBOT_MAX_GROUND_SPEED
        
        // Узнаем положение робота и мяча без высоты 
        let me_pos = 
            positionRobot me
            |> projectionXZ
        let ball_pos = 
            positionBall game.ball
            |> projectionXZ

        let target_delta = sub interception me_pos
        let target_velocity = divScalar target_delta max_speed

        let delta = sub ball_pos me_pos

        let dist_for_touch = rules.ROBOT_RADIUS + rules.BALL_RADIUS
        let max_jump_speed = rules.ROBOT_MAX_JUMP_SPEED
        // Если робот касается мяча и находится ниже его,
        // то прыгаем как можно сильнее, чтобы пнуть его.
        let jump_speed = 
            if length delta <= dist_for_touch && me.y < game.ball.y then 
                max_jump_speed
            else
                 0.0

        {
            target_velocity_x = vX target_velocity
            target_velocity_y = 0.0
            target_velocity_z = vZ target_velocity
            jump_speed = jump_speed
            use_nitro = true
        }


    let attackEntryPoint (me, rules, game, _) : Action option = 
        // Расстояние между мячом и роботом.
        let delta = sub (positionBall game.ball) (positionRobot me)

        let dist_for_touch = rules.ROBOT_RADIUS + rules.BALL_RADIUS
        let max_jump_speed = rules.ROBOT_MAX_JUMP_SPEED
        // Если робот касается мяча и находится ниже его,
        // то прыгаем как можно сильнее, чтобы пнуть его.
        let jump_speed = 
            if length delta <= dist_for_touch && me.y < game.ball.y then 
                max_jump_speed
            else
                 0.0
        
        // Узнаем положение робота и мяча и скорость мяча без высоты 
        let me_pos = 
            positionRobot me
            |> projectionXZ
        let ball_pos =
            positionBall game.ball
            |> projectionXZ
        let ball_vel =
            velocityBall game.ball
            |> projectionXZ

        // Провекрка возможности требуемой скорости
        let max_speed = rules.ROBOT_MAX_GROUND_SPEED
        let isPossibleSpeed speed = 
            0.5 * max_speed < speed && speed < max_speed

        // Узнаем положение мяча через определенное время
        let delta_ball_pos = mulScalar ball_vel 0.1
        let ballPosWithTime t  = 
            let time = float t * 0.1
            let new_delta_ball_pos = mulScalar delta_ball_pos time
            let new_pos = add ball_pos new_delta_ball_pos
            (new_pos, time)

        let ballInArena (ball_pos : Vector, _) = 
            abs(vX ball_pos) < rules.arena.width / 2.0 
            && abs(vZ ball_pos) < rules.arena.depth / 2.0

        // Для текущего положение мяча без высоты и времени
        // пытаемся узнать, сможем ли мы добежать до мяча чтобы пнуть его
        let tryAttack (ball_pos : Vector, time : float) = 
            // расстояние, требуемая скорость и вектор требуемой скорости
            let delta2 = sub ball_pos me_pos
            let need_speed = length delta2 / time
            let velocity = divScalar delta2 time
            if isPossibleSpeed need_speed then
                Some {                
                    target_velocity_x = vX velocity
                    target_velocity_y = vY velocity
                    target_velocity_z = vZ velocity
                    jump_speed = jump_speed
                    use_nitro = true
                }
            else
                None  

        // Перебираем положение мяча в будущем, 
        // до тех пор пока не выйдет за границы арены,
        // беря первый попавшийся вариант, в котором можем сможем добежать до мяча.
        Array.init 100 ballPosWithTime
        |> Array.takeWhile ballInArena
        |> Array.tryPick tryAttack


    let attackAct args : Action =
        // Если нашли коим образом можно атаковать,
        // то возвращаем это действие,
        // иначе переходим в режим защиты
        match attackEntryPoint args with
        | Some x -> x
        | None -> protectAct args


    let private tuple2 x y = (x, y)


    let nextData game : StrategyData = 
        let cur_tick = game.current_tick
        let tick = 
            match data with
            | KnownAttackBot(x, _) -> x 
        // Если информация актуальна то ничего не делаем,
        // иначе атакующим будет ближайший к воротам
        if cur_tick = tick then
            data
        else
            game.robots
            |> Array.filter (fun x -> x.is_teammate)
            |> Array.maxBy (fun x -> x.z)
            |> (fun x -> x.id)
            |> tuple2 cur_tick
            |> KnownAttackBot


    let assignFieldsAction toActoin fromAction = 
        toActoin.target_velocity_x <- fromAction.target_velocity_x
        toActoin.target_velocity_y <- fromAction.target_velocity_y
        toActoin.target_velocity_z <- fromAction.target_velocity_z
        toActoin.jump_speed <- fromAction.jump_speed
        toActoin.use_nitro <- fromAction.use_nitro

    let act (me : Robot, rules : Rules, game : Game, action : Action) =
        // Обновляем в случае необходимости id атакующего робота.
        data <- nextData game
        // Проверяем атакующий ли сейчас робот 
        let idAttackBot =
            match data with
            | KnownAttackBot(_, x) -> x 
        let newAction =         
            if me.id = idAttackBot then
                attackAct (me, rules, game, action)
            else
                protectAct (me, rules, game, action)
        assignFieldsAction action newAction

File: QUICKSTARTS/csharp/MyStrategy.cs
using System;
using Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk.Model;
using Action = Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk.Model.Action;

namespace Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk
{
    // Логика стратегии скопирована из базовой стратегии c++ (SannikovDmitry)
    public sealed class MyStrategy : IStrategy
    {
        private const double EPS = 1e-5;
        
        public void Act(Robot me, Rules rules, Game game, Action action)
        {
            var ball = game.ball;

            // Наша стратегия умеет играть только на земле
            // Поэтому, если мы не касаемся земли, будет использовать нитро
            // чтобы как можно быстрее попасть обратно на землю
            if (!me.touch)
            {
                action.target_velocity_x = 0.0;
                action.target_velocity_z = 0.0;
                action.target_velocity_y = -rules.MAX_ENTITY_SPEED;
                action.jump_speed = 0.0;
                action.use_nitro = true;
                return;
            }

            // Если при прыжке произойдет столкновение с мячом, и мы находимся
            // с той же стороны от мяча, что и наши ворота, прыгнем, тем самым
            // ударив по мячу сильнее в сторону противника
            bool jump = (DistBetweenBallAndRobot3D(ball, me) < (rules.BALL_RADIUS + rules.ROBOT_MAX_RADIUS) &&
                         me.y < ball.y);

            // Так как роботов несколько, определим нашу роль - защитник, или нападающий
            // Нападающим будем в том случае, если есть дружественный робот,
            // находящийся ближе к нашим воротам
            bool is_attacker = false;
            foreach (var robot in game.robots)
            {
                if (robot.is_teammate && robot.id != me.id)
                {
                    if (robot.z < me.z)
                    {
                        is_attacker = true;
                    }
                }
            }

            if (is_attacker)
            {
                // Стратегия нападающего:
                // Просимулирем примерное положение мяча в следующие 10 секунд, с точностью 0.1 секунда
                for (int i = 0; i < 100; ++i)
                {
                    double t = i * 0.1;
                    Ball ball_pos = CalcNewBallPos(ball, t);
                    // Если мяч не вылетит за пределы арены
                    // (произойдет столкновение со стеной, которое мы не рассматриваем),
                    // и при этом мяч будет находится ближе к вражеским воротам, чем робот,
                    if (ball_pos.z > me.z
                        && Math.Abs(ball.x) < (rules.arena.width / 2.0)
                        && Math.Abs(ball.z) < (rules.arena.depth / 2.0))
                    {
                        // Посчитаем, с какой скоростью робот должен бежать,
                        // Чтобы прийти туда же, где будет мяч, в то же самое время
                        double delta_pos_x = ball_pos.x - me.x;
                        double delta_pos_z = ball_pos.z - me.z;
                        double delta_pos_dist = Math.Sqrt(delta_pos_x*delta_pos_x + delta_pos_z*delta_pos_z);
                        double need_speed = delta_pos_dist / t;
                        // Если эта скорость лежит в допустимом отрезке
                        if (0.5 * rules.ROBOT_MAX_GROUND_SPEED < need_speed
                            && need_speed < rules.ROBOT_MAX_GROUND_SPEED)
                        {
                            // То это и будет наше текущее действие
                            action.target_velocity_x = delta_pos_x / delta_pos_dist * need_speed;
                            action.target_velocity_z = delta_pos_z / delta_pos_dist * need_speed;
                            action.target_velocity_y = 0.0;
                            action.jump_speed = jump ? rules.ROBOT_MAX_JUMP_SPEED : 0.0;
                            action.use_nitro = false;
                            return;
                        }
                    }
                }
            }
            
            // Стратегия защитника (или атакующего, не нашедшего хорошего момента для удара):
            // Будем стоять посередине наших ворот
            double target_pos_x = 0.0;
            double target_pos_z = -(rules.arena.depth / 2.0) + rules.arena.bottom_radius;
            // Причем, если мяч движется в сторону наших ворот
            if (ball.velocity_z < -EPS) {
                // Найдем время и место, в котором мяч пересечет линию ворот
                double t = (target_pos_z - ball.z) / ball.velocity_z;
                double x = ball.x + ball.velocity_x * t;
                // Если это место - внутри ворот
                if (Math.Abs(x) < (rules.arena.goal_width / 2.0)) {
                    // То пойдем защищать его
                    target_pos_x = x;
                }
            }
            
            // Установка нужных полей для желаемого действия
            double target_velocity_x = (target_pos_x - me.x)*rules.ROBOT_MAX_GROUND_SPEED;
            double target_velocity_z = (target_pos_z - me.z)*rules.ROBOT_MAX_GROUND_SPEED;
            
            action.target_velocity_x = target_velocity_x;
            action.target_velocity_z = target_velocity_z;
            action.target_velocity_y = 0.0;
            action.jump_speed = jump ? rules.ROBOT_MAX_JUMP_SPEED : 0.0;
            action.use_nitro = false;
        }

        public string CustomRendering()
        {
            return "";
        }
        
        private static double DistBetweenBallAndRobot3D(Ball b, Robot r)
        {
            return Math.Sqrt((b.x - r.x) * (b.x - r.x) + (b.y - r.y) * (b.y - r.y) + (b.z - r.z) * (b.z - r.z));
        }

        private static double DistBetweenBallAndRobot2D(Ball b, Robot r)
        {
            return Math.Sqrt((b.x - r.x) * (b.x - r.x) + (b.z - r.z) * (b.z - r.z));
        }

        private static Ball CalcNewBallPos(Ball b, double t)
        {
            return new Ball() {x = b.x + b.velocity_x * t, y = b.y + b.velocity_y * t, z = b.z + b.velocity_z * t};
        }
    }
}
